pykern/pkunit.py

```diff
@@ -184,88 +186,86 @@ def file_eq(expect_path, *args, **kwargs):
+            is_jinja = expect_path.ext == '.jinja'
+            b = expect_path.purebasename if is_jinja else expect_path.relto(data_dir())
+            if actual_path is None:
+                actual_path = b
+            if not isinstance(actual_path, pykern.pkconst.PY_PATH_LOCAL_TYPE):
+                actual_path = work_dir().join(actual_path)
+            self.expect_path = expect_path
+            self.actual_path = actual_path
+            self.is_jinja = is_jinja
+
+        def _xlsx_to_csv_convert(self, actual_xlsx, sheet):
+            try:
+                import pandas
+            except ModuleNotFoundError:
+                pkfail('optional module=pandas must be installed to compare xlsx={}', actual_xlsx)
+
+            p = pandas.read_excel(
+                actual_xlsx,
+                index_col=None,
+                sheet_name=sheet,
+            )
+            p.columns = p.columns.map(lambda c: '' if 'Unnamed' in str(c) else str(c))
+            p.to_csv(
+                str(self.actual_path),
+                encoding='utf-8',
+                index=False,
+                line_terminator='\r\n',
             )
 
-    def _xlsx_to_csv(params):
-        from pykern.pkdebug import pkdlog
-
-        try:
-            b = params.actual.new(ext='.xlsx')
-            m = _CSV_SHEET_ID.search(b.purebasename)
-            s = 0
-            if m:
-                b = b.new(purebasename=m.group(1))
-                s = int(m.group(2))
-            if b.check(file=True):
-                _xlsx_to_csv_convert(pkcollections.PKDict(
-                    actual_xlsx=b,
-                    sheet=s,
-                    actual_csv=params.actual
-                ))
-                # no xlsx to convert so just let file_eq handle normally
-                return
-        except Exception:
-            pkdlog('ERROR converting xlsx to csv expect={} actual={}', params.expect, params.actual)
-            raise
+        def _xlsx_to_csv(self):
+            from pykern.pkdebug import pkdlog
+
+            try:
+                b = self.actual_path.new(ext='.xlsx')
+                m = _CSV_SHEET_ID.search(b.purebasename)
+                s = 0
+                if m:
+                    b = b.new(purebasename=m.group(1))
+                    s = int(m.group(2))
+                if b.check(file=True):
+                    self._xlsx_to_csv_convert(b, s)
+                    # no xlsx to convert so just let file_eq handle normally
+                    return
+            except Exception:
+                pkdlog('ERROR converting xlsx to csv expect={} actual={}', self.expect_path, self.actual_path)
+                raise
+
+        def _convert_xlsx(self):
+            if self.expect_path.ext == '.csv' and not self.actual_path.check(file=True):
+                self._xlsx_to_csv()
+
+        def _set_expect_and_actual(self, actual_in_args, actual):
+            if self.expect_path.ext == '.json' and not self.actual_path.exists():
```

Comment:
dispatch to a function. This is really _convert_expect_and_actual

Reply:
Not sure what you mean. _set_expect_and_actual does not perform the conversion.